apps/emqx/src/emqx_persistent_session_ds.erl

```diff
@@ -394,7 +421,24 @@ handle_timeout(_ClientInfo, get_streams, Session) ->
 -spec replay(clientinfo(), [], session()) ->
     {ok, replies(), session()}.
 replay(_ClientInfo, [], Session = #{inflight := Inflight0}) ->
-    {Replies, Inflight} = emqx_persistent_message_ds_replayer:replay(Inflight0),
+    AckedUntil = emqx_persistent_message_ds_replayer:committed_until(ack, Inflight0),
+    RecUntil = emqx_persistent_message_ds_replayer:committed_until(rec, Inflight0),
+    CompUntil = emqx_persistent_message_ds_replayer:committed_until(comp, Inflight0),
+    ReplyFun = fun
+        (_Seqno, #message{qos = ?QOS_0}) ->
+            {false, []};
+        (Seqno, #message{qos = ?QOS_1}) when Seqno < AckedUntil ->
```

Comment:
(if the results of `committed_until` are inclusive) shouldn't this be `=<`?

Reply:
They are indeed exclusive. I.e. "committed until 42" means the last committed seqno is 41.